### PR TITLE
refactor: decouple router2 setup and creation logic from the config structs, switch to `Arc<dyn ServerType>`

### PR DESCRIPTION
**Pull request by alamb**
# Rationale:

re https://github.com/influxdata/influxdb_iox/issues/3926

I want to be able to create a router2 as part of an all in one server without having to explicitly enumerate out out all the concrete types (which are several levels deep and I don't think are required for performance).

Or put another way "I gave up trying to figure out all the generics to make this work" -- will annotate this PR with more details

# Changes:
1. Move the creation of the `RouterServerType` into the `router2` module in  influxdb_ioxd 
2. Switch to using `dyn ServerType` rather than generics over `<ServerType>`

The overhead of one extra dynamic dispatch for each http request seems pretty minimal, so I don't think this will have any measurable performance impact.

For a big picture view, you can see how this is used in the WIP PR for the all in one server https://github.com/influxdata/influxdb_iox/pull/3965

### CODE REVIEW

**Comment by alamb** [comment]
I think this should be uncontroversial, but want to make sure @crepererum  and @domodwyer are cool with it prior to merge

File: influxdb_iox/src/commands/run/router2.rs [hunk 1]
```diff
@@ -91,164 +80,15 @@ pub async fn command(config: Config) -> Result<()> {
         .get_catalog("router2", Arc::clone(&metrics))
         .await?;
 
-    // Initialise the sharded write buffer and instrument it with DML handler
-    // metrics.
-    let write_buffer = init_write_buffer(
-        &config,
+    info!("starting router2");
+    let server_type = create_router2_server_type(
```
[the remark this answers is not in the document]
Comment (carols10cents) [reply influxdb_iox/src/commands/run/router2.rs:79]: Previously, this code did all the initializing and then called `info!("starting router2");`. Now, the code calls `info!("starting router2");` and then does all the initializing-- was that intentional? Does it matter? I'm not sure, but I noticed the change and wanted to check :)

**Review by carols10cents — APPROVED** [review]
Looks good to me, just one thing I noticed that might be nothing!

Out of curiosity, did switching from a generic to a trait object have any noticeable effect on compile times?

**Comment by alamb** [comment]
> Previously, this code did all the initializing and then called info!("starting router2");. Now, the code calls info!("starting router2"); and then does all the initializing-- was that intentional? Does it matter? I'm not sure, but I noticed the change and wanted to check :)

It was not intentional -- good catch -- I will put it back

File: influxdb_iox/src/commands/run/router2.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (alamb) [reply influxdb_iox/src/commands/run/router2.rs:78]: fixed in 690d12f2a (it was not intentional)

**Review by crepererum — APPROVED** [review]

**Review by domodwyer — APPROVED** [review]
Nice!

**Comment by domodwyer** [comment]
Is it easy to keep the creation of the `RouterServerType` in the `rotuer2` package? Then we could use this new init method for the e2e test/benchmark (in the `router2` crate) too which would be much nicer...

**Comment by alamb** [comment]
> Is it easy to keep the creation of the RouterServerType in the rotuer2 package? Then we could use this new init method for the e2e test/benchmark (in the router2 crate) too which would be much nicer than what it does today.

Right now, all the tonic instantiation in influxdb_ioxd package.  I'll see if I can move more of this code into `router2` as well

**Pull request merged**